[MemProf] Support cloning for indirect calls with ThinLTO

teresajohnson opened the pull request:
This patch enables support for cloning in indirect callsites.

This is done by synthesizing callsite records for each virtual call
target from the profile metadata. In the thin link all the synthesized
records for a particular indirect callsite initially share the same
context node, but support is added to partition the callsites and
outgoing edges based on the callee function, creating a separate node
for each target.

In the LTO backend, when cloning is needed we first perform indirect
call promotion, then change the target of the new direct call to the
desired clone.

Note this is ThinLTO-specific, since for regular LTO indirect call
promotion should have already occurred.

teresajohnson committed:
[MemProf] Support cloning for indirect calls with ThinLTO

This patch enables support for cloning in indirect callsites.

This is done by synthesizing callsite records for each virtual call
target from the profile metadata. In the thin link all the synthesized
records for a particular indirect callsite initially share the same
context node, but support is added to partition the callsites and
outgoing edges based on the callee function, creating a separate node
for each target.

In the LTO backend, when cloning is needed we first perform indirect
call promotion, then change the target of the new direct call to the
desired clone.

Note this is ThinLTO-specific, since for regular LTO indirect call
promotion should have already occurred.

teresajohnson committed:
Test the -enable-memprof-indirect-call-support flag

teresajohnson commented:
I had forgotten to test the new internal option that can be used to disable the support, just added that.

teresajohnson committed:
Add missing check for null function (indirect call) when applying
cloning decisions.

snehasish commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4108,6 +4456,119 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+        // possible that this profile could have been collected on a different
+        // target (or version of the code), and we need to be conservative
+        // (similar to what is done in the ICP pass).
+        Function *TargetFunction = Symtab.getFunction(Candidate.Value);
```
Can this be const?

snehasish commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4108,6 +4456,119 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+        // Handle each call clone, applying ICP so that each clone directly
+        // calls the specified callee clone, guarded by the appropriate ICP
+        // check.
+        auto CalleeOrigName = TargetFunction->getName();
```
I think this is only used on L4537, can we inline this call there?

snehasish commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4108,6 +4456,119 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+        // calls the specified callee clone, guarded by the appropriate ICP
+        // check.
+        auto CalleeOrigName = TargetFunction->getName();
+        for (unsigned J = 0; J < NumClones; J++) {
+          CallBase *CBClone;
+          // Copy 0 is the original function.
+          if (!J)
+            CBClone = CB;
+          else
+            CBClone = cast<CallBase>((*VMaps[J - 1])[CB]);
```
nit: Could be a little more compact --
```
for (unsigned J = 0, CallBase *CBClone = CB; J < NumClones; J++) {
  if (J > 0)
    CBClone = cast<CallBase>((*VMaps[J - 1])[CB]);
  ....
}
```

snehasish commented on llvm/include/llvm/Transforms/IPO/MemProfContextDisambiguation.h:
```diff
@@ -43,8 +43,13 @@ class MemProfContextDisambiguation
 
+  // Whether we are building with SamplePGO. This is needed for correctly
+  // updating profile metadata on speculatively promoted calls.
+  bool SamplePGO;
```
nit: IsSamplePGO

snehasish commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4108,6 +4456,119 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+          auto *TargetToUse = TargetFunction;
+          // Call original if this version calls the original version of its
+          // callee.
+          if (StackNode.Clones[J])
```
Braces around this if condition would make it easier to read.

snehasish commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4108,6 +4456,119 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+      // Adjust the MD.prof metadata for all clones, now that we have the new
+      // TotalCount and the number promoted.
+      for (unsigned J = 0; J < NumClones; J++) {
+        CallBase *CBClone;
+        // Copy 0 is the original function.
+        if (!J)
+          CBClone = CB;
+        else
+          CBClone = cast<CallBase>((*VMaps[J - 1])[CB]);
```
Same simplification as the one suggested above?

snehasish commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4075,24 +4367,80 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
-        } else if (CB->isTailCall()) {
+            CloneCallsite(StackNode, CB, CalledFunction);
+          }
+        } else if (CB->isTailCall() && CalledFunction) {
```
Is it necessary to check CalledFunction here? I believe a tail call should always be direct and thus we should always have CalledFunction.

Maybe an assert is more appropriate if you agree.

snehasish commented on llvm/test/ThinLTO/X86/memprof-icp.ll:
```diff
@@ -0,0 +1,320 @@
+;; using llvm-reduce with the expected FileCheck input.
+
+;; -- virtfunc.h: --
+;; #include <unistd.h>
```
nit: This header is needed in virtfunc_main.cc not here.

snehasish commented on llvm/test/ThinLTO/X86/memprof-icp.ll:
```diff
@@ -0,0 +1,320 @@
+;; promoted call to the correct clone.
+
+;; This was created from the following source code, then the IR was reduced
+;; using llvm-reduce with the expected FileCheck input.
```
I think we should also mention the step where the profile is collected and applied to the IR (which was then reduced)? I see the metadata has InstrProf so we should mention that here too.

@kazutakahirata I think here a text format profile would have helped significantly. The other profiles mixed in already have text formats so just because memprof doesn't it makes constructing the test even more tedious.

snehasish commented on llvm/test/ThinLTO/X86/memprof-icp.ll:
```diff
@@ -0,0 +1,320 @@
+
+!0 = !{i32 1, !"ProfileSummary", !1}
+!1 = !{!2, !3, !4, !5, !6, !7, !8, !9, !10, !11}
+!2 = !{!"ProfileFormat", !"InstrProf"}
```
Is it worth adding a test for SampleProf? 

Maybe add TODO somewhere in the code to add a test once we have a text format to prevent any changes from breaking our usage.

teresajohnson reviewed (commented):
I've addressed most of the comments except of the refactoring suggestions, which I will do in a follow on update most likely tomorrow.

teresajohnson committed:
Address most of the review comments, remaining are TODO

teresajohnson commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4108,6 +4456,119 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+    // version of the target, which is what is in the vtable.
+    for (auto &ICallInfo : ICallAnalysisMap) {
+      auto *CB = ICallInfo.first;
+      auto &Info = ICallInfo.second;
```
actually this one became moot when I changed it to a vector

teresajohnson committed:
Remove a couple of changes only needed with the structured binding that
was no longer applicable with the change from a map to a vector.

teresajohnson commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4075,24 +4367,80 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
-          assert(SI != FS->callsites().end());
-          auto &StackNode = *(SI++);
-
+          if (!CalledFunction) {
```
done on moving this handling to a helper. I made ICallAnalysis a private member, however it is initialized by another helper at the beginning of applyImport, since it isn't needed in other cases that use this class (e.g. fullLTO). I didn't move ICallAnalysisMap (now ICallAnalysisInfo) since that is a per-function data structure.

teresajohnson commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4108,6 +4456,119 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+    // devirtualization) for each clone of the callsite, and update its callee
+    // to the appropriate clone. Note that the ICP compares against the original
+    // version of the target, which is what is in the vtable.
+    for (auto &ICallInfo : ICallAnalysisMap) {
```
done

teresajohnson committed:
Address remaining comments

teresajohnson commented:
> I've addressed most of the comments except of the refactoring suggestions, which I will do in a follow on update most likely tomorrow.

I've finished addressing all the comments

snehasish reviewed (approved):
lgtm

teresajohnson commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4457,121 +4425,172 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+
+void MemProfContextDisambiguation::performICP(
+    Module &M, ArrayRef<CallsiteInfo> AllCallsites,
+    SmallVectorImpl<std::unique_ptr<ValueToValueMapTy>> &VMaps,
+    SmallVector<ICallAnalysisData> &ICallAnalysisInfo,
```
done

teresajohnson commented on llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:
```diff
@@ -4457,121 +4425,172 @@ bool MemProfContextDisambiguation::applyImport(Module &M) {
+      const char *Reason = nullptr;
+      if (!isLegalToPromote(*CB, TargetFunction, &Reason)) {
+        ORE.emit([&]() {
+          return OptimizationRemarkMissed(DEBUG_TYPE, "UnableToPromote", CB)
```
done

teresajohnson committed:
Address review comments
